### PR TITLE
BICO-3457 Added the custom field action for shopping list line items

### PR DESCRIPTION
**Pull request by avareldev**

### CODE REVIEW

File: src/ActionBuilder/ShoppingList/SetLineItemCustomField.php [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+ * @package BestIt\CommercetoolsODM\ActionBuilder\ShoppingList
+ */
+class SetLineItemCustomField extends ShoppingListActionBuilder
+{
+    /**
+     * A PCRE to match the hierarchical field path without delimiter.
+     *
+     * @var string
+     */
+    protected $complexFieldFilter = 'lineItems/([^/]+)';
```
[the remark this answers is not in the document]
Comment (b3nl) [reply src/ActionBuilder/ShoppingList/SetLineItemCustomField.php]: But you just want to map a single line item ...

[the remark this answers is not in the document]
Comment (avareldev) [reply src/ActionBuilder/ShoppingList/SetLineItemCustomField.php]: Changed to lineItems/([^/]+)/custom

File: tests/ActionBuilder/ShopplingList/SetLineItemCustomFieldTest.php [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+     *
+     * The First Element is the field path, the second element is the reference class and the optional third value
+     * indicates the return value of the support method.
+     *
+     * @return array
+     */
+    public function getSupportAssertions(): array
+    {
+        return [
+            ['lineItems/', LineItem::class, true],
```
[the remark this answers is not in the document]
Comment (avareldev) [reply tests/ActionBuilder/ShopplingList/SetLineItemCustomFieldTest.php]: Done.

File: src/ActionBuilder/ShoppingList/SetLineItemCustomField.php [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+use Commercetools\Core\Request\AbstractAction;
+use Commercetools\Core\Request\ShoppingLists\Command\ShoppingListSetLineItemCustomTypeAction;
+
+/**
+ * Builds the action to change cart item custom fields
+ *
+ * @author André Varelmann <andre.varelmann@bestit-online.de>
+ * @package BestIt\CommercetoolsODM\ActionBuilder\ShoppingList
+ */
+class SetLineItemCustomField extends ShoppingListActionBuilder
```
[the remark this answers is not in the document]
Comment (avareldev) [reply src/ActionBuilder/ShoppingList/SetLineItemCustomField.php]: Changed.

File: tests/ActionBuilder/ShopplingList/SetLineItemCustomFieldTest.php [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+/**
+ * Test for the shopping list set line item custom field action
+ *
+ * @author André Varelmann <andre.varelmann@bestit-online.de>
+ * @package BestIt\CommercetoolsODM\Tests\ActionBuilder\ShopplingList
+ */
+class SetLineItemCustomFieldTest extends TestCase
+{
+    /**
+     * @var SetLineItemCustomField|void The tested class.
```
[the remark this answers is not in the document]
Comment (avareldev) [reply tests/ActionBuilder/ShopplingList/SetLineItemCustomFieldTest.php]: Changed to |null

File: src/ActionBuilder/ShoppingList/SetLineItemCustomField.php [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        string $subFieldName = ''
+    ): array {
+        $actions = [];
+
+        $offset = $this->getLastFoundMatch()[1];
+        $lineItemId = $sourceObject->getLineItems()->getAt($offset)->getId();
+
+        // Only process if custom fields were changed and the line item already exists
+        if ($lineItemId !== null && isset($changedValue['custom']['fields'])) {
+            $actions[] = $this->createCustomTypeAction($lineItemId, $changedValue['custom']);
```
[the remark this answers is not in the document]
Comment (avareldev) [reply src/ActionBuilder/ShoppingList/SetLineItemCustomField.php]: I think it's okay sind a type without fields has no use.

**Pull request merged**